File: analysis/nationalextract.R
setwd("~/research/tippt/META/analysis/")

library(dplyr)

alldf <- data.frame()
for (rcp in c("1.5-Base", "CP-Base", "NP-Base")) { # "RCP2.6", "RCP4.5", "RCP6",
    print(rcp)
    df1 <- read.csv(paste0("../results/bytimexcountry-SSP2-", rcp, "-NoOMH-temps.csv"))
    df0 <- read.csv(paste0("../results/bytimexcountry-SSP2-", rcp, "-NoTPs-temps.csv"))

    df1.dmg <- read.csv(paste0("../results/bytimexcountry-SSP2-", rcp, "-NoOMH-damages.csv"))
    df0.dmg <- read.csv(paste0("../results/bytimexcountry-SSP2-", rcp, "-NoTPs-damages.csv"))

    df1.gwl <- read.csv(paste0("../results/bytime-SSP2-", rcp, "-NoOMH-gwarm.csv"))
    df0.gwl <- read.csv(paste0("../results/bytime-SSP2-", rcp, "-NoTPs-gwarm.csv"))

    df <- df0 %>% left_join(df1, by=c('time', 'country', 'trialnum'), suffix=c('.no', '.tp')) %>%
        left_join(df0.dmg, by=c('time', 'country', 'trialnum')) %>%
        left_join(df1.dmg, by=c('time', 'country', 'trialnum'), suffix=c('.no', '.tp')) %>%
        left_join(df0.gwl, by=c('time', 'trialnum')) %>%
        left_join(df1.gwl, by=c('time', 'trialnum'), suffix=c('.no', '.tp'))

    alldf <- rbind(alldf, cbind(rcp=rcp, df %>% filter(time %% 5 == 0)))
}

alldf2 <- alldf %>% filter(time %in% c(2025, 2035, 2050, 2100)) %>%
    group_by(rcp, country, time) %>%
    summarize(GWL.no=mean(T_AT.no, na.rm=T), GWL.tpdiff=mean(T_AT.tp - T_AT.no, na.rm=T),
              T.no=mean(T_country.no, na.rm=T), T.tpdiff=mean(T_country.tp - T_country.no, na.rm=T),
              D.no=mean(total_damages_percap_peryear_percent.no, na.rm=T),
              D.tpdiff=mean(total_damages_percap_peryear_percent.tp -
                            total_damages_percap_peryear_percent.no, na.rm=T))

as.data.frame(subset(alldf2, country == 'USA' & time == 2025))

## For dataset
dataset <- rbind(alldf %>% filter(time %in% c(2025, 2035, 2050) & rcp == "RCP4.5"),
                 alldf %>% filter(time %in% c(2050) & rcp == "RCP2.6"),
                 alldf %>% filter(time == 2100 & rcp %in% c("RCP2.6", "RCP4.5", "RCP6")))
save(dataset, file="META-NoOMH-compare.RData")

write.csv(subset(dataset, country == "USA"), "META-NoOMH-NCA6.csv", row.names=F)

File: analysis/nationalextract.jl
using Mimi
import Random
include("../src/MimiMETA.jl")
include("../src/montecarlo.jl")
include("../src/bge.jl")

# Scenarios

for rcp in ["RCP4.5", "RCP3-PD/2.6", "RCP6", "NP-Base", "CP-Base", "1.5-Base"]
    for TP in ["NoTPs", "NoOMH"]
        if TP == "NoOMH"
            global model = full_model(;
                                      rcp = rcp, # Concatenate correct scenario-variant name
                                      ssp = "SSP2",
                                      #co2 = "Expectation",
                                      #ch4 = "default",
                                      #warming = "Best fit multi-model mean",
                                      tdamage = "pointestimate",
                                      slrdamage = "mode",
                                      saf = "Distribution mean",
                                      interaction = true,
                                      pcf = "Fit of Hope and Schaefer (2016)",
                                      omh = false,
                                      amaz = "Cai et al. central value",
                                      gis = "Nordhaus central value",
                                      ais = "AIS",
                                      ism = "Value",
                                      amoc = "IPSL",
                                      nonmarketdamage = true)
        else
            global model = full_model(;
                                      rcp = rcp, # Concatenate correct scenario-variant name
                                      ssp = "SSP2",
                                      #co2 = "Expectation",
                                      #ch4 = "default",
                                      #warming = "Best fit multi-model mean",
                                      tdamage = "pointestimate",
                                      slrdamage = "mode",
                                      saf = "Distribution mean",
                                      interaction = false,
                                      pcf = false,
                                      omh = false,
                                      amaz = false,
                                      gis = false,
                                      ais = false,
                                      ism = false,
                                      amoc = false,
                                      nonmarketdamage = true)
        end

        println("$rcp SSP2 $TP")

        myupdate_param!(model, :Consumption, :damagepersist, 0.25)

        Random.seed!(26052023)

        ### Run the model so we can run scripts
        run(model)

        ### Run the model in MC mode
        if TP == "NoOMH"
            results = sim_full(model, 1000,
                               "Fit of Hope and Schaefer (2016)", # PCF
                               "Cai et al. central value", # AMAZ
                               "Nordhaus central value", # GIS
                               "none", # WAIS
                               "Distribution", # SAF
                               true, # ais_used
                               true, # ism_used
                               false, # omh_used
                               true, # amoc_used
                               false, # persist
                               false, # emuc
                               false; # prtp
                               save_rvs=false)
        else
            results = sim_full(model, 1000,
                               "none", # PCF
                               "none", # AMAZ
                               "none", # GIS
                               "none", # WAIS
                               "Distribution", # SAF
                               false, # ais_used
                               false, # ism_used
                               false, # omh_used
                               false, # amoc_used
                               false, # persist
                               false, # emuc
                               false; # prtp
                               save_rvs=false)
        end
        run(model) # model is overwritten in some cases

        if rcp == "RCP3-PD/2.6"
            rcpfile = "RCP2.6"
        else
            rcpfile = rcp
        end

        #Export country-level damages
        df = simdataframe(model, results, :TotalDamages, :total_damages_percap_peryear_percent)
        CSV.write("../results/bytimexcountry-SSP2-$rcpfile-$TP-damages.csv", df[(df.time .>= 2010) .& (df.time .<= 2100), :])

        #Export country-level temperatures
        df = simdataframe(model, results, :PatternScaling, :T_country)
        CSV.write("../results/bytimexcountry-SSP2-$rcpfile-$TP-temps.csv", df[(df.time .>= 2010) .& (df.time .<= 2100), :])

        df = simdataframe(model, results, :TemperatureConverter, :T_AT)
        CSV.write("../results/bytime-SSP2-$rcpfile-$TP-gwarm.csv", df[(df.time .>= 2010) .& (df.time .<= 2100), :])
    end
end
